### DISCUSSION TITLE
[犀牛鸟-E1]：结项报告-即插即用的 ES-MoE 块插件 At PyPI

### DISCUSSION THREAD

**Discussion by Lfan-ke** [Tencent/YOLO-Master#276] [Show and tell]
E1 课题结题：一个装在官方 Ultralytics 上的 ES-MoE 块插件，以及与本仓的同配置对照实验。

**目录**：[完成程度](#完成程度) · [发布模型复核](#发布模型复核) · [同配置对照](#同配置对照) · [机制观察](#机制观察) · [已合入](#已合入) · [复现与文档](#复现与文档)

| | |
|:--:|:--:|
| 安装 | `pip install esmoe`（[PyPI](https://pypi.org/project/esmoe/) 1.0.1） |
| 仓库 | https://github.com/Lfan-ke/ES-MoE |
| 文档 | https://lfan-ke.github.io/ES-MoE/ |
| 实验与图表 | https://lfan-ke.github.io/ES-MoE/experiments/ |
| 与 YOLO-Master 逐项对照 | https://lfan-ke.github.io/ES-MoE/design/ |
| 记录与权重 | [results/](https://github.com/Lfan-ke/ES-MoE/tree/main/results)、[`checkpoints` 分支](https://github.com/Lfan-ke/ES-MoE/tree/checkpoints) |
| 判读线 | https://lfan-ke.github.io/ES-MoE/JUDGMENT/ |

```python
import esmoe
model = esmoe.equip("yolo11n.yaml", weight=0.01)
model.train(data="visdrone.yaml", epochs=120, imgsz=800)
```

## 完成程度

任务书把 E1 分为 P0、P1、P2 三档，P2+ 指超出 P2 的部分。逐档对应如下：

| 档位 | 任务书要求 | 交付 |
|:--:|:--:|:--:|
| P0 保底 | 在统一预算下复核候选 ES-MoE 模块并选型，保留多 seed 证据 | [选型](https://lfan-ke.github.io/ES-MoE/SELECTION/)：2 / 4 / 8 专家与 top-1、top-2 在统一预算下比较，出厂配置 4 专家 top-2、权重 0.01 在全量 VisDrone、20 epoch、640 像素下三个 seed 确认（配对 3/3，mAP50 +0.0021），协议预算下 v8n +0.0025；七代主干三臂矩阵每格三个 seed 起 |
| P1 预期 | 完成 pip / 仓库插件化接入，兼容至少三代 Ultralytics 主干，提供 README 与 Colab | `pip install esmoe`，`equip` 一个调用接入；兼容 YOLOv5、v8、v9、v10、11、12、26 七代官方主干与 YOLO-Master 分支；README 与 [Colab 快速上手](https://colab.research.google.com/github/Lfan-ke/ES-MoE/blob/main/notebooks/quickstart.ipynb) |
| P2 理想 | 正式发布、产出中英文教程，并根据首轮社区反馈迭代 | PyPI 与 GitHub Release 1.0.1；文档站中英双语、按版本切换（教程、API、设计、实验、选型、判读线、发布说明）；反馈与迭代逐条写在[发布说明](https://lfan-ke.github.io/ES-MoE/RELEASE/) |
| P2+ 超出部分 | 任务书之外 | 与本仓分支同配置对照两轮，块的作用等效；发布模型复核到小数点后五位一致；142 次满协议运行、600 卡时，权重与逐 epoch 曲线全部公开；八轮判读线先于结果提交（11 次登记）；335 项测试，含与上游、论文的数值对照；死专家机制与平衡项梯度分析；上游修复 #241 已合入 |

任务书列出的交付物（插件包、兼容矩阵、多 seed 消融、README 与 Colab、发布说明）由仓库脚本 `scripts/closure.py` 逐项核对，结果在 [`results/closure.md`](https://github.com/Lfan-ke/ES-MoE/blob/main/results/closure.md)，未满足项为 0。

## 发布模型复核

YOLO-Master-EsMoE-N 的权重装进本包的块，在官方 Ultralytics 8.4.101 上跑 COCO val2017，与本仓分支到小数点后五位一致，参数量两边同为 2,694,364：

| | precision | recall | mAP50 | mAP50-95 |
|:--:|:--:|:--:|:--:|:--:|
| YOLO-Master 分支 | 0.68425 | 0.53357 | 0.58532 | 0.42288 |
| 官方 + esmoe | 0.68425 | 0.53357 | 0.58532 | 0.42288 |

该权重是每块 3 专家、top-3 全开，与仓库 yaml 在当前代码下建出的 4 专家 top-2 不同。

## 同配置对照

VisDrone 全量、imgsz 800、120 epoch、batch 32、`patience=0`，三个 seed，每个 seed 的四臂同卡：A 为本仓分支跑 `yolo-master-n`（四个 `ES_MOE`），A0 为同分支去块，B 为官方 8.4.101 加本包四个 `ESMoE` 并复刻本仓对路由模型的三项处理（辅助损失按滑动平均归一并封顶 3.0、路由器组半学习率且不进 Muon、前 3 epoch 冻专家），C 为官方去块。A 与 B 逐层参数一致（10 类时 2,814,406）。

共两轮：第七轮各臂按框架默认精度训练；第八轮四臂统一 FP32（`--amp 0`），排除训练精度不同的影响。两轮不在同一台机器上，之间只比方向。指标由 `scripts/measure.py` 从各 run 的 `last.pt` 统一重测，验证参数相同。

| 差（mAP50） | 第七轮 | 为正 | 第八轮（FP32） | 为正 |
|:--:|:--:|:--:|:--:|:--:|
| A − A0 | +0.0122 [+0.0019, +0.0225] | 3/3 | +0.0104 [+0.0015, +0.0193] | 3/3 |
| B − C | +0.0126 [+0.0070, +0.0181] | 3/3 | +0.0095 [−0.0047, +0.0237] | 3/3 |
| (A − A0) − (B − C) | −0.0004 [−0.0105, +0.0098] | 2/3 | +0.0009 [−0.0221, +0.0239] | 2/3 |
| A − B | −0.0069 [−0.0175, +0.0037] | 0/3 | +0.0046 [−0.0071, +0.0164] | 2/3 |

方括号内为 95% 区间。块在两个框架里的作用两轮都等效（差之差落在同配置重复运行的噪声底 0.0045 之内）；A − B 两轮都跨在预登记的两条判读线之间，记为无法判定，第八轮符号为正。第八轮 B − C 三个 seed 都为正、mAP50-95 也为正，按判读线有效；区间跨零，按噪声底三档无法判定。FP32 下同卡重复运行五对，平均差 0.0038、最大 0.0103，与混合精度的 0.0045 / 0.0130 同一量级，两套噪声底给出的档位相同。

面积分档（COCO 式）两轮都是大目标涨得最多，但大目标只有 2/3 个 seed 为正。逐 seed 的四臂、四个差的区间、训练耗时与重复运行都画成了交互式图表：[实验 · 同配置对照](https://lfan-ke.github.io/ES-MoE/experiments/#同配置对照)。

## 机制观察

平衡项读按 top-K 重归一后的门控，对未进 top-K 的专家梯度恒为零。读门控的目标下 6 个检查点里 5 个有死专家（进入 top-2 的验证图不足 1%），读完整 softmax 的 Switch 式 66 个里 0 个。两轮 B 臂的六个检查点都是四块里三块有死专家、块 3 都没有，路由概率的熵为最大值的 98.5%（第八轮 98.8%）。

论文式 (12)(13) 与代码的 GShard 式只差仿射 `L_论文 = (L_代码 − 1)/E²`，最小值点相同、梯度成正比。

## 已合入

`OptimizedMOE` 的追踪守卫只挡 ONNX、未挡 `torch.jit.trace`：[#241](https://github.com/Tencent/YOLO-Master/pull/241)。

## 复现与文档

```bash
git clone https://github.com/Lfan-ke/ES-MoE && cd ES-MoE
uv sync --group dev
uv run python scripts/same_config.py
uv run python scripts/routing.py --summarise
uv run python scripts/charts.py
```

分档与路由只读检查点。实验口径、数据集统计与流程图见[实验](https://lfan-ke.github.io/ES-MoE/experiments/)，块的结构与本仓 `ES_MOE` 的逐项对照见[ES-MoE 与 YOLO](https://lfan-ke.github.io/ES-MoE/design/)。

**Comment by Lfan-ke** [comment]
上一次双周周报：[位置](https://github.com/Lfan-ke/ES-MoE/issues/5)

**Comment by Lfan-ke** [comment]
相关链接：

- https://lfan-ke.github.io/ES-MoE/
- https://github.com/Lfan-ke/ES-MoE/wiki

> 完成程度： `E1` の `P2+`.
